README.md
# Selami Router

[![Build Status](https://api.travis-ci.org/selamiphp/router.svg?branch=master)](https://travis-ci.org/selamiphp/router) [![Coverage Status](https://coveralls.io/repos/github/selamiphp/router/badge.svg?branch=master)](https://coveralls.io/github/selamiphp/router?branch=master) [![Latest Stable Version](https://poser.pugx.org/selami/router/v/stable)](https://packagist.org/packages/selami/router) [![Total Downloads](https://poser.pugx.org/selami/router/downloads)](https://packagist.org/packages/selami/router) [![Latest Unstable Version](https://poser.pugx.org/selami/router/v/unstable)](https://packagist.org/packages/selami/router) [![License](https://poser.pugx.org/selami/router/license)](https://packagist.org/packages/selami/router)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/selamiphp/router/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/selamiphp/router/) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/748983d7d23e4c26b13dd76fc781cdc8)](https://www.codacy.com/app/mehmet/framework?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=selamiphp/router&amp;utm_campaign=Badge_Grade)


Router and Dispatcher built on top of [nikic/FastRoute](https://github.com/nikic/FastRoute) that returns controller class name to be instantiated, desired content-type and route aliases if defined.


## Installation

Install this library using composer:

```bash
$ composer require selami/router
```
## Example
Get global values.

```php
<?php
declare(strict_types=1);

$request = new PSR7ServerRequest(); // Let's say this class implements PSR7 ServerRequestInterface
$response = new PSR7ServerResponse(); // Let's say this class implements PSR7 ResponseInterface
$defaultReturnType  = 'json';       // Possible values: html, json, text, redirect, download. To be used to send output.
$requestMethod      = 'GET';        // i.e. $_SERVER['REQUEST_METHOD']
$requestedUri       = '/user/12/inbox'; // i.e. $_SERVER['REQUEST_URI']

```
Create Selami\Router Instance.

```php
$router = new Selami\Router(
    $defaultReturnType,
    $requestMethod,
    $requestedUri
);
```
Add routes that expect HTTP request methods. $route variable uses nikic/FastRoute's route syntax.

```php
$route = '/';
$action = Controllers\Home::class;
$returnType = 'html';
$alias = 'home';

$router->get(
    $route,         // required
    $action,        // required
    $returnType,    // optional, default: $defaultReturnType
    $alias          // optional, default null
);

$router->post('/login', Controllers\Login::class, 'redirect');
$router->get('/dashboard', Controllers\Dashboard::class, 'html', 'dashboard');
$router->get('/api/user/{id}', Controllers\Api\Users::class, 'json');
$router->get('/user/{id:\d+}/{box}', Controllers\Api\Users\Inbox::class, 'html', 'user_home');

```
Get requested route info and aliases.

```php

$routeInfo = $router->getRoute();

/*
$routeInfo = [
    'route' => [
        'controller' => "Controllers\Api\Users\Inbox"
        'returnType' => "html"
        'args' => [
            'id' => 12,
            'box' => "inbox"
         ]
         'status' => 200
    ]
    'aliases' => [
        'home' => "/"
        'dashboard' => "/dashboard"
    ]
];
 
*/

```
Now you can call your controller.

```php
$controller = new $routeInfo['route']['controller']($routeInfo['route']['args']);

$outputMethod = 'return' . ucfirst($routeInfo['route']['returnType']);

echo $controller->$outputMethod($request, $response);

```

Our sample Controller class can be like:
```php
<?php

declare(strict_types=1);

namespace Controller\Api\Users;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use BaseController;

class Inbox extends BaseController {
    
    private $args;
    
    public function __construct($args) 
    {
        parent::__construct();
        $this->args = $args;
    }
    
    public function returnHTML(ServerRequestInterface $request, ResponseInterface $response)
    {
        $response = $this->response->withHeader('Content-Type', 'text/html');
        $response->getBody()->write('Your user id is: ' . $this->args['id'] . '. You are viewing your '. $this->args['box']);
        return $response->output();
    }
}

```

## $route value syntax when adding get, post, put etc routes.

See [FastRoute documentation](https://github.com/nikic/FastRoute).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 0.2.0 - 2017-01-03

### Added

- __call magic method added to let router set http request methods using method call.(i.e. $router->get($route, $action)
- More exceptions defined
- Tests added/changed for changes

### Deprecated

- Nothing.

### Removed

- defined routes removed from constructor which breaks BC.

### Fixed

- Nothing.



test/bootstrap.php
<?php
namespace tests;
require_once dirname(__DIR__) . '/vendor/autoload.php';
$loader = new \Composer\Autoload\ClassLoader();
$loader->add('test', dirname(__DIR__) );
$loader->register();


test/RouterTest.php
<?php

namespace tests;

use Selami;
use Zend\Diactoros\ServerRequestFactory;
use ReflectionObject;
use UnexpectedValueException;
use InvalidArgumentException;

class MyRouterClass extends \PHPUnit_Framework_TestCase
{
    private $config = [
        'folder'                => '',
        'default_return_type'   =>'html'
    ];

    private $request;

    public function setUp()
    {
        $basedir = dirname(__DIR__) . '/app';
        $this->config['base_dir']   = $basedir;
        $this->config['app_dir']    = $basedir;
        $_SERVER                    = [];
        $_FILES                     = [];
        $_GET                       = [];
        $_POST                      = [];
        $_SERVER                    = [];
        $_COOKIE                    = [];
        $_SERVER['DOCUMENT_ROOT']   = $basedir . '/htdocs';
        $_SERVER['SCRIPT_NAME']     = '/index.php';
        $_SERVER['REQUEST_URI']     = '/alias';
        $_SERVER['SERVER_PROTOCOL'] = 'HTTP/1.1';
        $_SERVER['SERVER_NAME']     = 'Selami';
        $_SERVER['SERVER_PORT']     = '8080';
        $_SERVER['REQUEST_METHOD']  = 'GET';
        $_SERVER['QUERY_STRING']    = 'p1=1&p2=2';
        $_SERVER['HTTPS']           = '';
        $_SERVER['REMOTE_ADDR']     = '127.0.0.1';
        $_SERVER['REQUEST_TIME']    = time();
        $this->request              = ServerRequestFactory::fromGlobals($_SERVER, $_GET, $_POST, $_COOKIE, $_FILES);
    }

    /**
     * @test
     * @dataProvider extractFolderDataProvider
     * @param $requestedPath string
     * @param $folder string
     * @param $expected string
     */
    public function shouldExtractRouteFromURLSuccessfully($requestedPath, $folder, $expected)
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
        $router->add('get', '/', 'app/main',null, 'home');
        $router->add('get', '/json', 'app/json', 'json');
        $router->add('post', '/json', 'app/redirect', 'redirect');
        $router->add('get', '/alias', 'app/alias', null, 'alias');
        $reflector = new ReflectionObject($router);
        $method = $reflector->getMethod('extractFolder');
        $method->setAccessible(true);
        $result = $method->invoke($router, $requestedPath, $folder );
        $this->assertEquals(
            $expected,
            $result,
            'extractFolder did not correctly extract the requested path for sub folder'
         );
    }

    public function extractFolderDataProvider()
    {
        return [
            ['/', '', '/'],
            ['', '', '/'],
            ['/admin/dashboard', 'admin', '/dashboard']
        ];
    }

    /**
     * @test
     */
    public function shouldCorrectlyInstantiateRouter()
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
        $router->add('get', '/', 'app/main',null, 'home');
        $router->add('get', '/json', 'app/json', 'json');
        $router->add('post', '/json', 'app/redirect', 'redirect');
        $router->add('get', '/alias', 'app/alias', null, 'alias');
        $this->assertInstanceOf('Selami\Router', $router);
        $this->assertAttributeContains(
            'GET',
            'method',
            $router,
            "Router didn't correctly return method as GET."
        );
    }

    /**
     * @test
     */
    public function shouldCorrectlyReturnRouteAndRouteAliases()
    {
        $router = new Selami\Router(
            'json',
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );

        $router->get('/', 'app/main', null, 'home');
        $router->get('/json', 'app/json', 'json');
        $router->get('/html', 'app/json');
        $router->post('/json', 'app/redirect', 'redirect');
        $router->get('/alias', 'app/alias', null, 'alias');
        $routeInfo = $router->getRoute();
        $this->assertArrayHasKey('aliases', $routeInfo, "Router didn't correctly return route data");

        $this->assertArrayHasKey('home', $routeInfo['aliases'], "Router didn't correctly return aliases");
        $this->assertArrayHasKey('alias', $routeInfo['aliases'], "Router didn't correctly return aliases");
        $this->assertEquals('/', $routeInfo['aliases']['home'], "Router didn't correctly return aliases");
        $this->assertEquals('/alias', $routeInfo['aliases']['alias'], "Router didn't correctly return aliases");
        $this->assertArrayHasKey('controller', $routeInfo['route'], "Router didn't correctly return route data");
        $this->assertEquals('app/alias', $routeInfo['route']['controller'], "Router didn't correctly return router data");
        $this->assertEquals('json', $routeInfo['route']['returnType'], "Router didn't correctly return router data");
    }

    /**
     * @test
     * @expectedException UnexpectedValueException
     */
    public function shouldThrowUnexpectedValueExceptionFor__callMethod()
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
        $router->nonAvalibleHTTPMethod('/', 'app/main',null, 'home');
    }

    /**
     * @test
     * @expectedException UnexpectedValueException
     */
    public function shouldThrowUnexpectedValueExceptionFor__constructorMethod()
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            'UNEXPECTEDVALUE',
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
    }

    /**
     * @test
     * @expectedException UnexpectedValueException
     */
    public function shouldThrowUnexpectedValueExceptionForAddMethod()
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
        $router->add('nonAvailableHTTPMethod','/', 'app/main',null, 'home');
    }

    /**
     * @test
     * @expectedException InvalidArgumentException
     */
    public function shouldThrowInvalidArgumentExceptionForAddMethodIfREquestMEthotIsNotStringOrArray()
    {
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );
        $router->add(200,'/', 'app/main',null, 'home');
    }


    /**
     * @test
     */
    public function shouldCorrectlyReturnMethodNotAllowed()
    {
        $_SERVER['REQUEST_METHOD'] = 'POST';
        $this->request = ServerRequestFactory::fromGlobals($_SERVER, $_GET, $_POST, $_COOKIE, $_FILES);
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );

        $router->add('get', '/', 'app/main',null, 'home');
        $router->add('get', '/json', 'app/json', 'json');
        $router->add('post', '/json', 'app/redirect', 'redirect');
        $router->add('get', '/alias', 'app/alias', null, 'alias');
        $routeInfo = $router->getRoute();
        $this->assertEquals('405', $routeInfo['route']['status'], "Router didn't correctly return Method Not Allowed");
    }

    /**
     * @test
     */
    public function shouldCorrectlyReturnNotFound()
    {
        $_SERVER['REQUEST_URI'] = '/notexists';
        $_SERVER['REQUEST_METHOD'] = 'POST';
        $this->request = ServerRequestFactory::fromGlobals($_SERVER, $_GET, $_POST, $_COOKIE, $_FILES);
        $router = new Selami\Router(
            $this->config['default_return_type'],
            $this->request->getMethod(),
            $this->request->getUri()->getPath(),
            $this->config['folder']
        );

        $router->add('get', '/', 'app/main',null, 'home');
        $router->add('get', '/json', 'app/json', 'json');
        $router->add('post', '/json', 'app/redirect', 'redirect');
        $router->add('get', '/alias', 'app/alias', null, 'alias');
        $routeInfo = $router->getRoute();
        $this->assertEquals('404', $routeInfo['route']['status'], "Router didn't correctly returnNot FOund");
    }
}


src/Router.php
<?php
/**
 * Selami Router
 * PHP version 7+
 *
 * @license https://github.com/selamiphp/router/blob/master/LICENSE (MIT License)
 * @link https://github.com/selamiphp/router
 * @package router
 * @category library
 */

declare(strict_types = 1);

namespace Selami;

use FastRoute;
use InvalidArgumentException;
use UnexpectedValueException;

/**
 * Router
 *
 * This class is responsible for registering route objects,
 * determining aliases if available and finding requested route
 */
final class Router
{
    /**
     * routes array to be registered.
     * Some routes may have aliases to be used in templating system
     * Route item can be defined using array key as an alias key.
     * @var array
     */
    private $routes = [];

    /**
     * aliases array to be registered.
     * Each route item is an array has items respectively : Request Method, Request Uri, Controller/Action, Return Type.
     * @var array
     */
    private $aliases = [];

    /**
     * HTTP request Method
     * @var string
     */
    private $method;

    /**
     * Request Uri
     * @var string
     */
    private $requestedPath;

    /**
     * Default return type if not noted in the $routes
     * @var string
     */
    private $defaultReturnType;


    /**
     * Translation array.
     * Make sures about return type.
     * @var array
     */
    private static $translations = [
        'h'     => 'html',
        'html'  => 'html',
        'r'     => 'redirect',
        'redirect' => 'redirect',
        'j'     => 'json',
        'json'  => 'json',
        't'     => 'text',
        'text'  => 'text',
        'd'     => 'download',
        'download'  => 'download'
    ];

    /**
     * Valid Request Methods array.
     * Make sures about requested methods.
     * @var array
     */
    private static $validRequestMethods = [
        'GET',
        'OPTIONS',
        'HEAD',
        'POST',
        'PUT',
        'DELETE',
        'PATCH'
    ];


    /**
     * Valid Request Methods array.
     * Make sures about return type.
     * Index 0 is also default value.
     * @var array
     */
    private static $validReturnTypes = [
        'html',
        'json',
        'text',
        'redirect',
        'download'
    ];

    /**
     * Router constructor.
     * Create new router.
     *
     * @param string $defaultReturnType
     * @param string $method
     * @param string $requestedPath
     * @param string $folder
     * @throws UnexpectedValueException
     */
    public function __construct(
        string $defaultReturnType,
        string $method,
        string $requestedPath,
        string $folder = ''
    ) {
        if (!in_array($method, self::$validRequestMethods, true)) {
            $message = sprintf('%s is not valid Http request method.', $method);
            throw new UnexpectedValueException($message);
        }
        $this->method   = $method;
        $this->requestedPath     = $this->extractFolder($requestedPath, $folder);
        $this->defaultReturnType = self::$translations[$defaultReturnType] ?? self::$validReturnTypes[0];
    }

    /**
     * Remove sub folder from requestedPath if defined
     * @param string $requestPath
     * @param string $folder
     * @return string
     */
    private function extractFolder(string $requestPath, string $folder)
    {
        if (!empty($folder)) {
            $requestPath = '/' . trim(preg_replace('#^/' . $folder . '#msi', '/', $requestPath), '/');
        }
        if ($requestPath === '') {
            $requestPath = '/';
        }
        return $requestPath;
    }

    /**
     * add route to routes list
     * @param string|array requestMethods
     * @param string $route
     * @param string $action
     * @param string $returnType
     * @param string $alias
     * @throws InvalidArgumentException
     * @throws UnexpectedValueException
     */
    public function add($requestMethods, string $route, string $action, string $returnType=null, string $alias=null)
    {
        $requestMethodsGiven = is_array($requestMethods) ? (array) $requestMethods : [0 => $requestMethods];
        $returnType = $this->checkReturnType($returnType);
        foreach ($requestMethodsGiven as $requestMethod) {
            $this->checkRequestMethodParameterType($requestMethod);
            $this->checkRequestMethodIsValid($requestMethod);
            if ($alias !== null) {
                $this->aliases[$alias] = $route;
            }
            $this->routes[] = [strtoupper($requestMethod), $route, $action, $returnType];
        }
    }

    /**
     * @param string $method
     * @param array $args
     * @throws UnexpectedValueException
     */
    public function __call(string $method, array $args)
    {

        $this->checkRequestMethodIsValid($method);
        $defaults = [
            null,
            null,
            $this->defaultReturnType,
            null
        ];
        list($route, $action, $returnType, $alias) = array_merge($args, $defaults);
        $this->add($method, $route, $action, $returnType, $alias);
    }

    /**
     * @param string|null $returnType
     * @return string
     */
    private function checkReturnType($returnType) {
        return $returnType === null ? $this->defaultReturnType : self::$validReturnTypes[$returnType] ?? $this->defaultReturnType;
    }

    /**
     * @param string $requestMethod
     * Checks if request method is valid
     * @throws UnexpectedValueException;
     */
    private function checkRequestMethodIsValid(string $requestMethod)
    {
        if (!in_array(strtoupper($requestMethod), self::$validRequestMethods, true)) {
            $message = sprintf('%s is not valid Http request method.', $requestMethod);
            throw new UnexpectedValueException($message);
        }
    }

    /**
     * @param $requestMethod
     * @throws InvalidArgumentException
     */
    private function checkRequestMethodParameterType($requestMethod)
    {
        $requestMethodParameterType = gettype($requestMethod);
        if (!in_array($requestMethodParameterType, ['array', 'string'], true)) {
            $message = sprintf(
                'Request method must be string or array but %s given.',
                $requestMethodParameterType);
            throw new InvalidArgumentException($message);
        }
    }

    /**
     * Dispatch against the provided HTTP method verb and URI.
     * @return array
     */
    private function dispatcher()
    {
        $options = [
            'routeParser'   => FastRoute\RouteParser\Std::class,
            'dataGenerator' => FastRoute\DataGenerator\GroupCountBased::class,
            'dispatcher'    => FastRoute\Dispatcher\GroupCountBased::class,
            'routeCollector' => FastRoute\RouteCollector::class,
        ];
        /** @var RouteCollector $routeCollector */
        $routeCollector = new $options['routeCollector'](
            new $options['routeParser'], new $options['dataGenerator']
        );
        $this->addRoutes($routeCollector);
        return new $options['dispatcher']($routeCollector->getData());
    }

    /**
     * Define Closures for all routes that returns controller info to be used.
     * @param FastRoute\RouteCollector $route
     */
    private function addRoutes(FastRoute\RouteCollector $route)
    {
        foreach ($this->routes as $definedRoute) {
            $definedRoute[3] = $definedRoute[3] ?? $this->defaultReturnType;
            $route->addRoute(strtoupper($definedRoute[0]), $definedRoute[1], function ($args) use ($definedRoute) {
                list(,,$controller, $returnType) = $definedRoute;
                $returnType = Router::$translations[$returnType] ?? $this->defaultReturnType;
                return  ['controller' => $controller, 'returnType'=> $returnType, 'args'=> $args];
            });
        }
    }



    /**
     * Get router data that includes route info and aliases
     */
    public function getRoute()
    {
        $dispatcher = $this->dispatcher();
        $routeInfo  = $dispatcher->dispatch($this->method, $this->requestedPath);
        $routerData = [
            'route'     => $this->runDispatcher($routeInfo),
            'aliases'   => $this->aliases
        ];
        return $routerData;
    }


    /**
     * Get route info for requested uri
     * @param array $routeInfo
     * @return array $routerData
     */
    private function runDispatcher(array $routeInfo)
    {
        $routeData = $this->getRouteData($routeInfo);
        $dispatchResults = [
            FastRoute\Dispatcher::METHOD_NOT_ALLOWED => [
                'status' => 405
            ],
            FastRoute\Dispatcher::FOUND => [
                'status'  => 200
            ],
            FastRoute\Dispatcher::NOT_FOUND => [
                'status' => 404
            ]
        ];
        return array_merge($routeData, $dispatchResults[$routeInfo[0]]);
    }

    /**
     * Get routeData according to dispatcher's results
     * @param array $routeInfo
     * @return array
     */
    private function getRouteData(array $routeInfo)
    {
        if ($routeInfo[0] === FastRoute\Dispatcher::FOUND) {
            list(, $handler, $vars) = $routeInfo;
            return $handler($vars);
        }
        return [
            'status'        => 200,
            'returnType'    => 'html',
            'definedRoute'  => null,
            'args'          => []
        ];
    }
}
